# Flights Data Exploration
## By Ifeoma Benson

## Introduction
> The dataset reports flights in the United States, including carriers, arrival and departure delays from 1987 to 2008. I gathered & analyze data for three years (2006, 2007 & 2008).  The total number of  observations are 16,984,354  and the following questions will be addressed:-

- what are the worst airports/states in terms of delayed and cancelled flights?
- what are the worst flights in terms of delayed and cancelled flights?
- what carrier is mostly used? 
- Best day of the week to travel ?
- what year encountered  high number of cancellation in flights
 


## Preliminary Wrangling


In [1]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

In [2]:
# suppress warnings from final output
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

> Load in your dataset and describe its properties through the questions below. Try and motivate your exploration goals through this section.


In [3]:
# load in the dataset into a pandas dataframe
 
df3 = pd.read_csv(r'C:\Users\user\Desktop\airport\2006.csv')
df4 = pd.read_csv(r'C:\Users\user\Desktop\airport\2007.csv')
df5 = pd.read_csv(r'C:\Users\user\Desktop\airport\2008.csv')

In [4]:
##check info of the dataset
lis = [df4,df5,df3]
for df in lis:
    print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7453215 entries, 0 to 7453214
Data columns (total 29 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Year               int64  
 1   Month              int64  
 2   DayofMonth         int64  
 3   DayOfWeek          int64  
 4   DepTime            float64
 5   CRSDepTime         int64  
 6   ArrTime            float64
 7   CRSArrTime         int64  
 8   UniqueCarrier      object 
 9   FlightNum          int64  
 10  TailNum            object 
 11  ActualElapsedTime  float64
 12  CRSElapsedTime     float64
 13  AirTime            float64
 14  ArrDelay           float64
 15  DepDelay           float64
 16  Origin             object 
 17  Dest               object 
 18  Distance           int64  
 19  TaxiIn             int64  
 20  TaxiOut            int64  
 21  Cancelled          int64  
 22  CancellationCode   object 
 23  Diverted           int64  
 24  CarrierDelay       int64  
 25  WeatherDelay      

In [5]:
#checking shape of dataframe
for df in lis:
    print(df.shape)

(7453215, 29)
(2389217, 29)
(7141922, 29)


In [6]:
#total number of row i am working with
a = 0
for df in lis:
    a+=df.shape[0]
print(a)

16984354


In [7]:
#read carrier into a dataframe
df6 = pd.read_csv(r'C:\Users\user\Desktop\airport\carriers.csv')
df6['Code'].nunique()

1490

In [8]:
#join 3 table together
joined =pd.merge(df3, df4,  how="outer")
joined =pd.merge(joined, df5,  how="outer")

joined = pd.merge(joined, df3, how="outer")
joined = pd.merge(joined, df4, how="outer")
joined = pd.merge(joined, df5, how="outer")

In [9]:
#show information of new table
joined.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 16984354 entries, 0 to 16984353
Data columns (total 29 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Year               int64  
 1   Month              int64  
 2   DayofMonth         int64  
 3   DayOfWeek          int64  
 4   DepTime            float64
 5   CRSDepTime         int64  
 6   ArrTime            float64
 7   CRSArrTime         int64  
 8   UniqueCarrier      object 
 9   FlightNum          int64  
 10  TailNum            object 
 11  ActualElapsedTime  float64
 12  CRSElapsedTime     float64
 13  AirTime            float64
 14  ArrDelay           float64
 15  DepDelay           float64
 16  Origin             object 
 17  Dest               object 
 18  Distance           int64  
 19  TaxiIn             float64
 20  TaxiOut            float64
 21  Cancelled          int64  
 22  CancellationCode   object 
 23  Diverted           int64  
 24  CarrierDelay       float64
 25  WeatherDelay    

In [10]:
#show shape of new table
joined.shape

(16984354, 29)

In [11]:
#show head of table
joined.head(3)

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0  2006      1          11          3    743.0         745   1024.0   
1  2006      1          11          3   1053.0        1053   1313.0   
2  2006      1          11          3   1915.0        1915   2110.0   

   CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
0        1018            US        343  ...   45.0     13.0          0   
1        1318            US        613  ...   27.0     19.0          0   
2        2133            US        617  ...    4.0     11.0          0   

   CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0               NaN         0           0.0          0.0      0.0   
1               NaN         0           0.0          0.0      0.0   
2               NaN         0           0.0          0.0      0.0   

   SecurityDelay  LateAircraftDelay  
0            0.0                0.0  
1            0.0                0.0  
2            0.0                0.0  

[3 rows x 29 columns]

In [12]:
#read airport data
df7 = pd.read_csv(r'C:\Users\user\Desktop\airport\airports.csv')
df7.iata 

0       00M
1       00R
2       00V
3       01G
4       01J
       ... 
3371    ZEF
3372    ZER
3373    ZPH
3374    ZUN
3375    ZZV
Name: iata, Length: 3376, dtype: object

In [13]:
#display information on carrier
df6

Code                    Description
0        02Q                  Titan Airways
1        04Q             Tradewind Aviation
2        05Q            Comlux Aviation, AG
3        06Q  Master Top Linhas Aereas Ltd.
4        07Q            Flair Airlines Ltd.
...      ...                            ...
1486      ZW    Air Wisconsin Airlines Corp
1487      ZX                   Air Georgian
1488  ZX (1)                     Airbc Ltd.
1489      ZY         Atlantic Gulf Airlines
1490     ZYZ           Skyway Aviation Inc.

[1491 rows x 2 columns]

i want to replace Uniquecarrier code to airline name  

In [14]:
#print uniquecarrier
print(joined.UniqueCarrier.nunique())
'''shows the dataset of uniqueCarrier and i can see 20 unique code'''

21


'shows the dataset of uniqueCarrier and i can see 20 unique code'

In [15]:
'''collect all index of unique code'''
uniquecode = joined.UniqueCarrier.value_counts().index

In [16]:
'''filter out code not in the list and reset index'''
airline= df6[df6.Code.isin(uniquecode).reset_index(drop=True)]

In [17]:
#checking the shape
airline.shape

(21, 2)

In [18]:
airline.Description

84                                 Pinnacle Airlines Inc.
100                                American Airlines Inc.
206                                   Aloha Airlines Inc.
221                                  Alaska Airlines Inc.
263                                       JetBlue Airways
377                            Continental Air Lines Inc.
441                                  Delta Air Lines Inc.
505                           Atlantic Southeast Airlines
517                                Frontier Airlines Inc.
537                           AirTran Airways Corporation
619                                Hawaiian Airlines Inc.
848                          American Eagle Airlines Inc.
920                               Northwest Airlines Inc.
936                                           Comair Inc.
949                                 Skywest Airlines Inc.
1292                               ATA Airlines d/b/a ATA
1297                                United Air Lines Inc.
1308    US Air

In [19]:
#getting only airline name in airline description and assign it to new column
airline['name'] = airline['Description'].apply(lambda x : x.split(' A')[0])

C:\Users\user\AppData\Local\Temp\ipykernel_17808\4098283649.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  airline['name'] = airline['Description'].apply(lambda x : x.split(' A')[0])


In [20]:
'''set index to code'''
airline = airline.set_index('Code')

In [21]:
#display airline table
airline

Description                name
Code                                                                       
9E                               Pinnacle Airlines Inc.            Pinnacle
AA                               American Airlines Inc.            American
AQ                                  Aloha Airlines Inc.               Aloha
AS                                 Alaska Airlines Inc.              Alaska
B6                                      JetBlue Airways             JetBlue
CO                           Continental Air Lines Inc.         Continental
DL                                 Delta Air Lines Inc.               Delta
EV                          Atlantic Southeast Airlines  Atlantic Southeast
F9                               Frontier Airlines Inc.            Frontier
FL                          AirTran Airways Corporation             AirTran
HA                               Hawaiian Airlines Inc.            Hawaiian
MQ                         American Eagle Airlines Inc.      American Eagle
NW                              Northwest Airlines Inc.           Northwest
OH                                          Comair Inc.         Comair Inc.
OO                                Skywest Airlines Inc.             Skywest
TZ                               ATA Airlines d/b/a ATA                 ATA
UA                                United Air Lines Inc.              United
US    US Airways Inc. (Merged with America West 9/05...                  US
WN                               Southwest Airlines Co.           Southwest
XE                             Expressjet Airlines Inc.          Expressjet
YV                                   Mesa Airlines Inc.                Mesa

In [22]:
print(joined.UniqueCarrier)
'''replace code with name '''

0           US
1           US
2           US
3           US
4           US
            ..
16984349    DL
16984350    DL
16984351    DL
16984352    DL
16984353    DL
Name: UniqueCarrier, Length: 16984354, dtype: object


'replace code with name '

In [23]:
#to assess key and value, need to create a dict
dict_name = dict(airline.name)

In [24]:
print(dict_name)
'''replace the joined.uniquecarrier code with vale '''

{'9E': 'Pinnacle', 'AA': 'American', 'AQ': 'Aloha', 'AS': 'Alaska', 'B6': 'JetBlue', 'CO': 'Continental', 'DL': 'Delta', 'EV': 'Atlantic Southeast', 'F9': 'Frontier', 'FL': 'AirTran', 'HA': 'Hawaiian', 'MQ': 'American Eagle', 'NW': 'Northwest', 'OH': 'Comair Inc.', 'OO': 'Skywest', 'TZ': 'ATA', 'UA': 'United', 'US': 'US', 'WN': 'Southwest', 'XE': 'Expressjet', 'YV': 'Mesa'}


'replace the joined.uniquecarrier code with vale '

In [25]:
#replace uniquecarrier code with carrier name 
joined['UniqueCarrier'] = joined.UniqueCarrier.replace(dict_name.keys(),dict_name.values())

In [26]:
'''confirm changes '''
joined.UniqueCarrier

0              US
1              US
2              US
3              US
4              US
            ...  
16984349    Delta
16984350    Delta
16984351    Delta
16984352    Delta
16984353    Delta
Name: UniqueCarrier, Length: 16984354, dtype: object

In [27]:
joined.head(3)

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0  2006      1          11          3    743.0         745   1024.0   
1  2006      1          11          3   1053.0        1053   1313.0   
2  2006      1          11          3   1915.0        1915   2110.0   

   CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
0        1018            US        343  ...   45.0     13.0          0   
1        1318            US        613  ...   27.0     19.0          0   
2        2133            US        617  ...    4.0     11.0          0   

   CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0               NaN         0           0.0          0.0      0.0   
1               NaN         0           0.0          0.0      0.0   
2               NaN         0           0.0          0.0      0.0   

   SecurityDelay  LateAircraftDelay  
0            0.0                0.0  
1            0.0                0.0  
2            0.0                0.0  

[3 rows x 29 columns]

In [28]:
 #converting column name to small letter
joined.columns = joined.columns.str.lower()

In [29]:
#drop columns not needed
cleaned = joined.drop(['dayofmonth','flightnum','taxiin','taxiout','tailnum','actualelapsedtime','crselapsedtime','distance'],axis=1)

In [30]:
#creating only rows with cancelled flight to reduce running memory
cleaned_cancel = cleaned[cleaned['cancelled'] == 1]
cleaned_cancel.shape[0]

347124

In [31]:
#filtering the rows from original data frame to see how many rows had depdelay time above one or equal to 1 from original data, since it didnt affect cancellation
cleaned_delay = cleaned.query('depdelay >=1 ')
cleaned_delay.shape[0]

6964015

In [32]:
cleaned.shape[0]

16984354

### What is the structure of your dataset?

In this project, i am using six(6) datasets for three(3) years 2006 and 2008.
There are  16,984,354 flight observations with 29 features in  total.

i have 3 sub-dataset:
> cleaned_cancel :all rows whose flight were cancelled with a total of 347,124 observation


> cleaned_delay  : all rows whose departure time were delayed by at least one minute with a total of 6,964,015


> cleaned : all rows from the 3 dataset, with unused columns dropped with a total of 16,984,353 number of observation 


### What is/are the main feature(s) of interest in your dataset?
 I am interested in knowing the most used carriers, destinations ,preferrred dayofweek  , timeofday, with lesser cancellation or delays and the top reason for cancellation distributed in years 

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?

Destination will help me ascertain what destination is frequently used and which destination record high number of cancellation or delay.  DayOfWeek, will answer my question of period with the least cancellation or delays. uniquecarrier will help me ascertain carrier with high number of cancellation or delay, same with cancellationcode and  crsdeptime. 

## Univariate Exploration
 


In [ ]:
#describe the joined table
cleaned.describe()

## What is the peak  Cancellationcode?

In [ ]:
#replacing all abbreviated letter to full name
cleaned.cancellationcode = cleaned.cancellationcode.replace({'A':'Carrier','B':'Weather','C':'Nas','D':'Security'})
base_color = sns.color_palette()[0]

In [ ]:
plt.figure(figsize=(14,7))
plot_cancel = cleaned.cancellationcode.value_counts()
plot_cancel_index = plot_cancel.index
plot_cancel_value = plot_cancel.values
sns.barplot(y= plot_cancel_index,x=plot_cancel_value, color=base_color)
plt.title('Distribution of cancellationcode')
plt.ylabel('Cancellationcode')
plt.xlabel('count');

This plot helped me displayed the distribution of cancellation cause. Carrier had the highest cancellation cause, followed by weather and Nasdelay , i will investigate it further in next section


 

### The Distribution of flights by dayofweek

In [ ]:
#create a label of day of the week
plt.figure(figsize=(10,7))
label = ['Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday']
#create a barplot 
sns.countplot(data=cleaned, x='dayofweek',color=base_color)#order=label
plt.xticks(rotation=15,ticks=[0,1,2,3,4,5,6],labels=label)
plt.ylabel('flights in million')
plt.xlabel('day of week')
plt.title('Distribution of cancelled flights by day of week');

This barplot shows that Weekend(saturday and sunday) had little drop in  flights during weekend 

#### Distribution of flights by uniquecarrier code

In [ ]:
#plot distribution of uniquecarrier
plt.figure(figsize=(13,7))
cleaned.uniquecarrier.value_counts().plot(kind = 'bar')
plt.title('distribution of uniquecarrier')
plt.ylabel('Flights in million')
plt.xlabel('UniqueCarrier')

i made this plot see which uniquecarrier is mostly used, i discovered Southwest was the highest number of flights followed by American and Skywest

### Distribution of Flight by Years

In [ ]:
#plot distribution of years using  value count
cleaned.year.value_counts().plot(kind= 'bar')
plt.title('Distribution of Flights by years ')
plt.xlabel('years')
plt.ylabel('frequency in Million')

This plot shows year 2008 had lower flight, i want to To investigate the drop in flight year 2008, i will queried the dataset below and find my answer. it also shows that year 2007 has the highest flight recorded.

In [ ]:
#used query to filter only months
abs = cleaned.query('year == 2008')
abs.month.value_counts()

After qeurying the year column in the dataset, i observe year 2008 had only 4 months recorded.

### Discuss the distribution(s) of your variable(s) of interest. Were there any unusual points? Did you need to perform any transformations?

> My variable of interest was CarrierDelay, UniqueCarriers, Day of the week, Cancellation Code and years distribution . I didnt perform any Transformation

### Of the features you investigated, were there any unusual distributions? Did you perform any operations on the data to tidy, adjust, or change the form of the data? If so, why did you do this?

>  The Annual distribution for flight usage had only 4 months recorded in year 2008 which reduced observation by 68%. i used numpy to change the abbrevations in uniquecarrier, dayofweek and cancellation code

## Bivariate Exploration
 

 

#### What is The Distribution of  Cancellation Code from 2006 to 2008?




In [ ]:
#used query method to filter dataset for year , cancellation and just the first quater

cancellation_2006 = cleaned.query('year == 2006 & month <=4')
cancellation_2007 = cleaned.query('year == 2007 & month <=4')
cancellation_2008 = cleaned.query('year == 2008 & month <=4')


plt.figure(figsize=(15,5))
plt.subplot(1,3,1)


#bar plot of cancellation cause in 2006
cancellation_2006.cancellationcode.value_counts().plot(kind='bar')
plt.xlabel('2006 cancellation cause')
plt.ylabel('Frequency')
plt.title(' cancellationcode first quater 2006')


plt.subplot(1,3,2)
#bar plot of cancellation cause in 2007
cancellation_2007.cancellationcode.value_counts().plot(kind='bar')
plt.xlabel('2007 cancellation cause')
plt.title(' cancellationcode first quater 2007')



plt.subplot(1,3,3)
#bar plot of cancellation cause in 2008
cancellation_2008.cancellationcode.value_counts().plot(kind='bar')
plt.xlabel('2008 cancellation cause')
plt.title(' cancellationcode first quater 2008')
plt.suptitle('Cancellation cause in first Quater through out years');


### I decided to use just first quater of each year because i want to show the equal distibution of cancellation code for 3 years, and its shows 2007 had high cancellation cause based on frequency and weather was the leading cause , 2008 had carrier and weather as leading cause while 2007 had carrier as leading cause


## How weather affected Flight cancellation?

In [ ]:
#group the weather delay in time into 5 and create new column
cleaned_cancel.loc[cleaned_cancel['weatherdelay'] <=15, 'weather_cat'] = 1
cleaned_cancel.loc[cleaned_cancel['weatherdelay'] >=15, 'weather_cat'] = 2
cleaned_cancel.loc[cleaned_cancel['weatherdelay'] >=60, 'weather_cat'] = 3
cleaned_cancel.loc[cleaned_cancel['weatherdelay'] >=120, 'weather_cat'] = 4
cleaned_cancel.loc[cleaned_cancel['weatherdelay'] >180, 'weather_cat'] = 5
cleaned_cancel.loc[cleaned_cancel['weatherdelay'].isna(), 'weather_cat'] =0 
 

In [ ]:
#group the weather delay in category that is associated to cancelled flight and assign to a new variable weather count
weather_count = cleaned_cancel.groupby(['weather_cat','cancelled'])['weather_cat'].size()
weather_count

#### Recall 0.0 means 'NAN' and 1.0 means less than 15 mins

In [ ]:
#taking percentage 
weather_count =  weather_count.reset_index(name='count')
weather_count['count_percentage'] = round((weather_count['count']/weather_count['count'].sum())*100,2)

 

In [ ]:
#show result
weather_count

In [ ]:
#percentage of cancelled flight from total
cancel_percent = round((weather_count['count'].sum()/cleaned.shape[0])*100,2)

In [ ]:
#replacing 0 and 1 with not_cancelled and cancelled respectively
weather_count['cancelled'] = weather_count.cancelled.replace({0:'not_Cancelled',1:'cancelled'})

In [ ]:
#replacing number with more meaningfull name 
weather_count.weather_cat = weather_count.weather_cat.replace({0.0:'NaN',1.0:' <= 15mins',2.0:'>=15mins',3.0:'>=60mins',4.0:'>=120mins',5.0:'>180mins'})


In [ ]:
#displaying weather_count table
weather_count

In [ ]:
#plot showing the cancelled flight due to weather delay
label =  weather_count.weather_cat 
plt.figure(figsize=(5,5))
plt.pie(weather_count.count_percentage,colors=['orange','brown'],labels=(label.values),autopct='%1.1f%%',pctdistance=.85)
plt.title('percentage of flight cancelled due to Weather delay',color='black')
circle = plt.Circle((0, 0), 0.7, color='white')
sum_ = weather_count.count_percentage.sum()
fig = plt.gcf()
plt.text(0., 0., cancel_percent, horizontalalignment='center', verticalalignment='center',fontsize=32)
fig.gca().add_artist(circle)
plt.rc('font', size=12)
plt.legend(loc=4)
;
plt.show();

i did this plot to show how much weather delay caused flight cancellation. A total of  2.04% accountted for cancelled flight which  were caused by weather delay with less than 15 mins with 81.4% and not-available minutes  for 18.6%

## Flights delayed due to weather?


In [ ]:
sns.scatterplot(data= cleaned_delay, x='weatherdelay',y='depdelay')
plt.title('Departure and Weather delay');

you can see above that depature delay associated with weather delay were between 0-16 minutes wait time 

### Day of the week that is likely for a flight to be cancelled 

In [ ]:
# replacing integer to readable column from only cancelled flight
cleaned_cancel.dayofweek = cleaned_cancel.dayofweek.replace({1:'Monday',2:'Tuesday',3:'Wednesday',4:'Thursday',5:'Friday',6:'Saturday',7:'Sunday'})

In [ ]:
#group dayofweek and cancelled 
daypivot = cleaned_cancel.groupby(['dayofweek','cancelled']).size()
dayfreq = daypivot.copy()

In [ ]:
#assign only index to daypivot
daypivot = daypivot.index.levels[0]
#assign values to dayfreq
dayfreq = dayfreq.values
 
#assign values to dayfreq
dayfreq

In [ ]:
#create a label of day of the week
label = ['Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday']
#create a barplot 
sns.barplot(y=daypivot, x=dayfreq,color=base_color,order=label)
plt.xticks(rotation=15)
plt.xlabel('Frequency')
plt.title('Distribution of cancelled flights by day of week');

This plot is to show Flights During the dayofweek that had  more cancellation and it appears weekend(saturday, sunday) had less cancellation, while friday had higher cancellation.
### Distribution of flight delay and day of the week.

In [ ]:
# replacing integer to readable column from only delayed flight
cleaned_delay.dayofweek = cleaned_delay.dayofweek.replace({1:'Monday',2:'Tuesday',3:'Wednesday',4:'Thursday',5:'Friday',6:'Saturday',7:'Sunday'})

In [ ]:
# make a plot of flights delay by dayofweek
cleaned_delay.dayofweek.value_counts().plot(kind='bar')
plt.title('Flight delayed by dayofweek')
plt.ylabel('frequency in millions')
plt.xlabel('day of week ')

This plot is to show Flights During the dayofweek that had  more delay and it appears weekend(saturday) had less delay, while friday experienced high delay with no much differnce accross the week.

### Which Carrier experienced high flight cancellation?

In [ ]:
#value count of uniquecarriers and assigning to variable called carrier_Value
carrier_delay = cleaned_cancel.uniquecarrier.value_counts()
carrier_value = carrier_delay.values
carrier_value

In [ ]:
#plot carrier associated with more cancelled flight

#create a function to plot distribution of delayed and cancelled flight by uniquecarrier 
def count_plot(x,y,z):
    fig,axes = plt.subplots(figsize=(15,10))
    #make a stem plot
    axes.stem(x,y, use_line_collection=True, basefmt=' ')
    # starting value of y-axis
    axes.set_ylim(0)
    plt.xticks(rotation = 90);
    plt.tight_layout()
    plt.title('Carrier associated with more  cancelled flight')
    plt.xlabel('Unique Carrier')
    plt.ylabel('FRequency');
    plt.title('Carrier associated with more {} flight'.format(z))
count_plot(carrier_delay.index,carrier_value,'cancelled')


#### the above plot is to show The Airline who had more flight cancellation, it Appears American Eagle has the highest flight cancellation through out the 3 years 

## which Airline has more depature delay time ?

In [ ]:
#value count of uniquecarriers and assigning to variable called carrier_Value
carrier_delay_2 = cleaned_delay.uniquecarrier.value_counts()
carrier_value_2 = carrier_delay_2.values
count_plot(carrier_delay_2.index,carrier_value_2,'Delayed')

#### The above plot is to show The Airline who had more flight DEpature Delay, it Appears Southwest has the highest flight DEpature delay through out the 3 years. Recall from univariate plot, Southeast was the most used carrier.

### What destination City are prone to Flight Cancellation?


In [ ]:
#use cancelled rows to search for destination that are prone to flight cancellation
cleaned_dest = cleaned_cancel.dest.value_counts().head(15)
dest_valu = cleaned_dest.values
 

#plot a barplot of the distribution above
plt.figure(figsize=(14,8))
plt.subplot(1,2,1)
sns.barplot(y=cleaned_dest.index , x=dest_valu,color=base_color)
plt.title('Distribution of flight cancelled and Destination ')
plt.ylabel('Destination')
plt.xlabel('count');

#frequency of most origin destination
cleaned_dest_1 = cleaned.dest.value_counts().head(15)
dest_valu_1 = cleaned_dest_1.values
#plot the distribution of flight from destination
plt.subplot(1,2,2)
sns.barplot(y=cleaned_dest_1.index , x=dest_valu_1, color=base_color)
plt.title('Distribution of flight from Destination ')
plt.ylabel('Destination')
plt.xlabel('count in million');

The left hand  plot shows the destination with high cancellation and ORD was the highest.  The right hand plot shows the destination with the highest record flight in total  and ATL has the highest destination.

### Which Origin experienced more  Flight Cancellation?

In [ ]:
#use cancelled rows to search for destination that are prone to flight cancellation
cleaned_origin = cleaned_cancel.origin.value_counts().head(15)
origin_valu = cleaned_origin.values
cleaned_dest

plt.figure(figsize=(14,7))

#plot a barplot of the distribution above
plt.subplot(1,2,1)
sns.barplot(y=cleaned_origin.index , x=origin_valu,color=base_color)
plt.title('Relationship of flight cancelled and Origin ')
plt.ylabel('Origin')
plt.xlabel('count');


#frequency of most origin destination
cleaned_origin_2 = cleaned.origin.value_counts().head(15)
origin_valu_2 = cleaned_origin_2.values

plt.subplot(1,2,2)
sns.barplot(y=cleaned_origin_2.index , x=origin_valu_2,color=base_color)
plt.title('Distribution of flight from Origin ')
plt.ylabel('Origin')
plt.xlabel('count in millions');

> The left hand  plot shows the origin with high cancellation and ORD was the highest.  The right hand plot shows the origin with the highest record flight in total  and ATL has the highest origin.

## What is the frequency of flight  cancelled or delayed by scheduledtime

In [ ]:
#plot frequency of flight occurence by time 
Schedhours= round((cleaned_cancel.crsdeptime/100),0)
Schedhours_1 = round((cleaned_delay.crsdeptime/100),0)
Schedhours_2 = round((cleaned.crsdeptime/100),0)

#plot frequency of flight cancelled by time 
plt.figure(figsize=(20,10))
plt.subplot(1,3,1)
Schedhours.value_counts(10).plot(kind='bar')
plt.title('distribution of flight cancelled by time')
plt.xlabel('Time')
plt.ylabel('count');

#plot frequency of flight delayed by time 
plt.subplot(1,3,2)
Schedhours_1.value_counts(10).plot(kind='bar')
plt.title('distribution of flight delayed by time')
plt.xlabel('Time')
plt.ylabel('count');

#plot frequency of flight  by time 
plt.subplot(1,3,3)
Schedhours_2.value_counts(10).plot(kind='bar')
plt.title('distribution of flights by time')
plt.xlabel('Time')
plt.ylabel('count');

> For cancelled Flight , scheduled flight had lowest average cancellation on saturday around 12pm - 1pm  and most cancelled flight on (monday,tuesday,thursday, friday) almost at same average time of 2pm - 4pm

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

- Destination with high cancellation and delay is ORD while most flight ORIGIN is ATL
- SouthWest carrier had the most depature delay while America Eagles had the most cancellation, southWest carrier was the most used Carrier. Overall, carrier accounted for 2.04% of flight cancellation
- saturday had low cancellation and delays.
- 6pm  has the highest number of flights with cancellation, with no delays. 4pm had a little high delay, and 8am is the highest frequent time for Flight 
 
### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

> No

## Multivariate Exploration

 

## Relationship between Depature Delay or Cancellation ,Scheduled Flight by Day of weeK?

In [ ]:
#plot Distribution of schedflight time that was delayed by dayofweek' 
plt.figure(figsize=(20,8))
plt.subplot(1,3,1);
Schedhours = cleaned_delay.crsdeptime/100

sns.boxplot(y= Schedhours, x='dayofweek' ,data= cleaned_delay,color= base_color, order=['Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday'])
plt.yticks(np.arange(0,24,2))
plt.title(' scheduledflight time that was delayed by dayofweek')
plt.tight_layout();

#plot scheduledflight time that was cancelled by dayofweek
plt.subplot(1,3,2);
Schedhours = cleaned_cancel.crsdeptime/100
sns.boxplot(y= Schedhours, x='dayofweek' ,data= cleaned_cancel,color= base_color)
plt.yticks(np.arange(0,24,2))
plt.xticks([0,1,2,3,4,5,6],['Mon','Tues','Wed','Thur','Fri','Sat','Sun'])
plt.title('scheduledflight time that was cancelled by dayofweek')
plt.tight_layout();


>For Delayed Flights, Saturday scheduled flight had the lowest chances of been delayed  with average scheduled time at 1pm - 2pm and sunday with the most delay flight around 3pm- 4pm .
> For cancelled Flights, Thursday had the lowest cancellation flights with an average time at 12:30pm - 1:30pm

### Talk about some of the relationships you observed in this part of the investigation. Were there features that strengthened each other in terms of looking at your feature(s) of interest?

6pm  has the highest number of flights with cancellation, with no delays. 4pm had a little high delay, and 8am is the highest frequent time for Flight      

### Were there any interesting or surprising interactions between features?

- Friday has high cancellations and delays.


## Conclusions
 
-  SouthWest Airlines Carrier operates high number of flights in U.S. with less cancellations but high delay while America Eagle had more cancellation

- ORD destination had the highest record flight cancellation and ATL has the highest Origin

In [ ]:
from subprocess import call
call(['python', '-m', 'nbconvert', 'exploration_template.ipynb'])

#saving Dataframe to csv file
cleaned.to_csv('cleaned.csv')
cleaned_delay.to_csv('cleaned_delay.csv')
cleaned_cancel.to_csv('cleaned_cancel.csv')